### PR TITLE
Added Artist search, Artist details, and network manager.

### PR DESCRIPTION
**Pull request by ZPendragon**

### CODE REVIEW

File: SpotifyArtistsViewer/AppDelegate.swift [hunk 1]
```diff
@@ -13,31 +13,14 @@ import CoreData
 class AppDelegate: UIResponder, UIApplicationDelegate {
 
     var window: UIWindow?
-
+   
 
     func application(application: UIApplication, didFinishLaunchingWithOptions launchOptions: [NSObject: AnyObject]?) -> Bool {
         // Override point for customization after application launch.
```
Comment (chow-land) [inline SpotifyArtistsViewer/AppDelegate.swift:18]: remove boilerplate comments

File: SpotifyArtistsViewer/SAArtist.swift [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+//  SAArtist.swift
+//  SpotifyArtistsViewer
+//
+//  Created by Kevin Zeckser on 6/6/16.
+//  Copyright © 2016 Kevin Zeckser. All rights reserved.
+//
+
+import UIKit
+
+struct SAArtist {
```
Comment (chow-land) [inline SpotifyArtistsViewer/SAArtist.swift]: just curious, why a struct over a class?

File: SpotifyArtistsViewer/SAArtist.swift [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    var image : String?
+    var description: String?
+    
+    init(name: String, image: String?, description: String?) {
+        self.name = name
+        self.image = image ?? "NoImageDefault"
+        self.description = description ?? "This band is awesome!"
+    }
+    
+}
```
Comment (chow-land) [inline SpotifyArtistsViewer/SAArtist.swift]: add linebreak at EOF (we try to do this across the board on projects)

File: SpotifyArtistsViewer/SAArtistViewController.swift [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    
+    @IBOutlet weak private var artistImageView: UIImageView!
+    @IBOutlet weak var descriptionLabel: UILabel!
+    var url :String?
+    var detailArtist: SAArtist?
+    
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        self.navigationController?.navigationBar.barTintColor = UIColor.blackColor()
```
Comment (chow-land) [inline SpotifyArtistsViewer/SAArtistViewController.swift]: removes self unless in a closure

Comment (chow-land) [inline SpotifyArtistsViewer/SAArtistViewController.swift]: Also, you can remove the 'UIColor' here, as the type is inferred from the left side. Keeps it a little cleaner.

File: SpotifyArtistsViewer/SAArtistViewController.swift [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+    // MARK: - Configure View Functions
+    
+    private func loadImage(url :String) -> UIImage? {
+        guard
+            let imgURL = NSURL(string: url),
+            let imgData = NSData(contentsOfURL: imgURL)
+            else { return nil }
+        return UIImage(data: imgData)
+
```
Comment (chow-land) [inline SpotifyArtistsViewer/SAArtistViewController.swift]: remove unnecessary whitespace

File: SpotifyArtistsViewer/SARequestManager.swift [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+//  Created by Kevin Zeckser on 6/6/16.
+//  Copyright © 2016 Kevin Zeckser. All rights reserved.
+//
+import Foundation
+
+enum SAResponse {
+    case Success(artists: [SAArtist])
+    case Failure(error: ErrorType)
+}
+
```
Comment (chow-land) [inline SpotifyArtistsViewer/SARequestManager.swift]: remove extra linebreak

File: SpotifyArtistsViewer/SARequestManager.swift [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+final class SARequestManager {
+    
+    // MARK: - Properties
+    
+    static let sharedInstance = SARequestManager()
+    private let session: NSURLSession!
+    // This prevents others from using the default '()' initializer for this class.
+    private init() {
+        let sessionConfig: NSURLSessionConfiguration = NSURLSessionConfiguration.defaultSessionConfiguration()
```
Comment (chow-land) [inline SpotifyArtistsViewer/SARequestManager.swift]: you don't need the ": NSURL... " type on the lefthand side here, as it's already inferred from the right side.

File: SpotifyArtistsViewer/SARequestManager.swift [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+        session = NSURLSession(configuration: sessionConfig)
+    }
+    
+    // MARK: - NSURLRequest & JSON Parse
+    
+    func getArtistsWithQuery(query: String, completion: SAResponse -> Void) {
+        let path = "https://api.spotify.com/v1/search?q=\(query)&type=track,artist&market=US"
```
Comment (chow-land) [inline SpotifyArtistsViewer/SARequestManager.swift]: make the URL path a constant property of the class.

File: SpotifyArtistsViewer/SARequestManager.swift [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+        guard let url = NSURL(string: path) else { return }
```
Comment (chow-land) [inline SpotifyArtistsViewer/SARequestManager.swift]: you can also have a property for the NSURL, which has a getter that does this check for you.

File: SpotifyArtistsViewer/SARequestManager.swift [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+        
+        let task = session.dataTaskWithURL(url) { (data: NSData?, response: NSURLResponse?, error: NSError?) in
+            var result: SAResponse
+            
+            // We have to handle a few cases here:
```
Comment (chow-land) [inline SpotifyArtistsViewer/SARequestManager.swift]: remove unhelpful comments

File: SpotifyArtistsViewer/SARequestManager.swift [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+            if let error = error {
+                // 1. Something went wrong with the request. We pass along the error.
+                result = SAResponse.Failure(error: error)
+            } else if let data = data {
+                
+                do {
+                    let jsonResult = try NSJSONSerialization.JSONObjectWithData(data, options: NSJSONReadingOptions.MutableContainers) as? NSDictionary
+                    var returnedArtists = [SAArtist]()
+                    
+                    if let artistResponse = jsonResult?["artists"], let artists = artistResponse["items"] as? NSArray {
```
Comment (chow-land) [inline SpotifyArtistsViewer/SARequestManager.swift]: I would put these Let's on separate lines, for readability. It's a little easy to miss the 2nd one.

File: SpotifyArtistsViewer/SARequestManager.swift [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+                        for entry in artists {
+                            let artistEntry = entry as? NSDictionary
+                            guard let name = artistEntry?["name"] as? String else { continue }
```
Comment (chow-land) [inline SpotifyArtistsViewer/SARequestManager.swift]: Typically the semantic meaning for a guard statement is "ensure this state is true before we carry out some logic".

In this case, I think it would be more intuitive if there was a guard simply to ensure that an artist entry existed, and then have a traditional if statement adding any fields that are available in the json. I think the only mission-critical thing for adding to the returnedArtists array is whether or not there is an artistEntry.

Also, are you sure this is the logic you want here? If there's an artist entry but they have no images, they should be skipped?

File: SpotifyArtistsViewer/SARequestManager.swift [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+                            guard let images = artistEntry?["images"] as? NSArray else { continue }
+                            guard let image = self.fetchImage(images as [AnyObject]) else { continue }
+                            
+                            // TODO: Load a description
+                            // let description: String?
```
Comment (chow-land) [inline SpotifyArtistsViewer/SARequestManager.swift]: TODO's are fine, but remove the commented code.

File: SpotifyArtistsViewer/SARequestManager.swift [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+                    }
+                
+                    result = SAResponse.Success(artists: returnedArtists)
+                } catch let error as NSError {
+                    result = SAResponse.Failure(error: error)
+                }
+            } else {
+                result = SAResponse.Success(artists: [SAArtist]())
+            }
+            // Call our completion handler on the main queue
```
Comment (chow-land) [inline SpotifyArtistsViewer/SARequestManager.swift]: remove unnecessary comment -- the code tells you what it's doing.

File: SpotifyArtistsViewer/SARequestManager.swift [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+            dispatch_async(dispatch_get_main_queue()) {
+                completion(result)
+            }
+        }
+        task.resume()
+
```
Comment (chow-land) [inline SpotifyArtistsViewer/SARequestManager.swift]: remove extra whitespace

File: SpotifyArtistsViewer/SARequestManager.swift [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+            }
+            // Call our completion handler on the main queue
+            dispatch_async(dispatch_get_main_queue()) {
+                completion(result)
+            }
+        }
+        task.resume()
+        
+    }
+
```
Comment (chow-land) [inline SpotifyArtistsViewer/SARequestManager.swift]: remove extra whitespace all day. Seems trivial but vertical space is precious when you're scrolling through a large file with other things open as well.

File: SpotifyArtistsViewer/SARequestManager.swift [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+    }
+ 
+    
+    func fetchImage(images: [AnyObject]?) -> String?  {
+        // take an array of strings, determine which is best for the job
+        guard let images = images else { return nil }
+        
+        for image in images {
+            
+            guard ((image as? [String:AnyObject]) != nil) else {
```
Comment (chow-land) [inline SpotifyArtistsViewer/SARequestManager.swift]: don't think you don't need the ' != nil ' part here

File: SpotifyArtistsViewer/SARequestManager.swift [hunk 18]
```diff
@@ -0,0 +1,101 @@
...
+                completion(result)
+            }
+        }
+        task.resume()
+        
+    }
+ 
+    
+    func fetchImage(images: [AnyObject]?) -> String?  {
+        // take an array of strings, determine which is best for the job
```
Comment (chow-land) [inline SpotifyArtistsViewer/SARequestManager.swift]: remove this comment, or make it helpful.

File: SpotifyArtistsViewer/SARequestManager.swift [hunk 19]
```diff
@@ -0,0 +1,101 @@
...
+        guard let images = images else { return nil }
+        
+        for image in images {
+            
+            guard ((image as? [String:AnyObject]) != nil) else {
+                return nil
+            }
+            
+            if let imageDict = image as? [String:AnyObject] {
```
Comment (chow-land) [inline SpotifyArtistsViewer/SARequestManager.swift]: the above guard statement should be  guard let imageDict = image as? [String,AnyObject]) else {... }. That way you already have this variable and can reduce nesting.

File: SpotifyArtistsViewer/SARequestManager.swift [hunk 20]
```diff
@@ -0,0 +1,101 @@
...
+                            return url
+                        }
+                    }
+                }
+            }
+
+        }
+        return nil
+    }
+}
```
Comment (chow-land) [inline SpotifyArtistsViewer/SARequestManager.swift]: linebreak

File: SpotifyArtistsViewer/SARequestManager.swift [hunk 21]
```diff
@@ -0,0 +1,101 @@
...
+        guard let images = images else { return nil }
+        
+        for image in images {
+            
+            guard ((image as? [String:AnyObject]) != nil) else {
+                return nil
+            }
+            
+            if let imageDict = image as? [String:AnyObject] {
+                if let height = imageDict["height"] as? Int, let width = imageDict["width"] as? Int {
```
Comment (chow-land) [inline SpotifyArtistsViewer/SARequestManager.swift]: can probably refactor this to reduce pyramidness

File: SpotifyArtistsViewer/SASearchViewController.swift [hunk 22]
```diff
@@ -0,0 +1,86 @@
...
+    
+    @IBOutlet weak var textField: UITextField!
+    @IBOutlet weak var searchSegmentControl: UISegmentedControl!
+    private var artists = [SAArtist]()
+    private let requestManager = SARequestManager.sharedInstance
+ 
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        self.navigationController?.navigationBar.barTintColor = UIColor.blackColor()
+        self.view.backgroundColor = UIColor.blackColor()
```
Comment (chow-land) [inline SpotifyArtistsViewer/SASearchViewController.swift]: remove selfs and 'UIColor's

File: SpotifyArtistsViewer/SASearchViewController.swift [hunk 23]
```diff
@@ -0,0 +1,86 @@
...
+            print("Success!!!")
+            self.artists = returnedArtists
+            self.tableView.reloadData()
+        }
+    }
+    
+    // MARK: - UITableViewDataSource
+
+    override internal func numberOfSectionsInTableView(tableView: UITableView) -> Int {
+
```
Comment (chow-land) [inline SpotifyArtistsViewer/SASearchViewController.swift]: remove whitespace

File: SpotifyArtistsViewer/SASearchViewController.swift [hunk 24]
```diff
@@ -0,0 +1,86 @@
...
+    
+    // MARK: - UITableViewDataSource
+
+    override internal func numberOfSectionsInTableView(tableView: UITableView) -> Int {
+        
+        return 1
+    }
+
+    override internal func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+
```
Comment (chow-land) [inline SpotifyArtistsViewer/SASearchViewController.swift]: same

File: SpotifyArtistsViewer/SASearchViewController.swift [hunk 25]
```diff
@@ -0,0 +1,86 @@
...
+        cell?.textLabel?.text = self.artists[indexPath.row].name
+        return cell!
+    }
+    
+    // MARK: - UITextFieldDelegate
+    
+    @objc internal func textFieldShouldReturn(textField: UITextField) -> Bool {
+        
+        guard let searchQuery = self.textField.text else { print("yall's text field dun broke"); return true }
```
Comment (chow-land) [inline SpotifyArtistsViewer/SASearchViewController.swift]: put this else block on a new line(s).  You should never have to use a semicolon in swift.

File: SpotifyArtistsViewer/SAArtist.swift (hunk 2 above)
Comment (ZPendragon) [inline SpotifyArtistsViewer/SAArtist.swift]: No particular reason. Initially I thought I'd take advantage of the default initializer from structs.
